muikku-core-plugins/src/main/resources/META-INF/resources/scripts/src/reducers/main-function/calendar.ts

```diff
@@ -0,0 +1,149 @@
+import { ActionType } from "~/actions";
+
+/**
+ *  Participant
+ */
+export type Participant = {
+  userEntityId: number;
+  name: string;
+  attendance: "UNCONFIRMED" | "YES" | "NO" | "MAYBE";
+};
+
+export type eventDisplay =
+  | "auto"
+  | "block"
+  | "list-item"
+  | "background"
+  | "inverse-background"
+  | "none";
+
+/**
+ *  CalendarEvent
+ */
+export interface CalendarEvent {
+  id?: number;
+  title?: string;
+  start: string;
+  end?: string;
+  /** Can the event overlap with others */
+  overlap?: boolean;
+  /** Override for the event */
+  editable?: boolean;
+  classNames?: string[];
+  description?: string;
+  /** How the event is displayed in the calendar */
+  display?: eventDisplay;
+  backgroundColor?: string;
+  /** Fullcalendar Resource */
+  resourceId?: string;
+  participants?: Participant[];
+}
+
+/**
+ *  CalendarEventStatusUpdate
+ */
+export interface CalendarEventStatusUpdate {
+  id: string;
+  status: "YES" | "NO" | "MAYBE";
+}
+
+/**
+ * EventState
+ */
+export type EventsState = "LOADING" | "ERROR" | "READY";
+
+/**
+ *  Calendar
+ */
+export interface Calendar {
+  state: EventsState;
+  guidanceEvents: CalendarEvent[];
+}
+
+/**
+ * Evaluates an event against an array of events
+ * @param event Event to be evaluated
+ * @param events Events to evaluate against
+ * @returns boolean
+ */
+export const evaluateEvents = (
+  event: CalendarEvent,
+  events: CalendarEvent[]
+) => {
+  let passed = true;
+  for (let i = 0; i < events.length; i++) {
+    if (events[i].id === event.id) {
+      passed = false;
+    }
+  }
+  return passed;
+};
+
+/**
+ * Reducer function for calendar
+ *
+ * @param state
+ * @param action
+ * @returns returns state
+ */
+export default function calendar(
+  state: Calendar = {
+    state: "LOADING",
+    guidanceEvents: [],
+  },
+  action: ActionType
+): Calendar {
+  switch (action.type) {
```

Comment:
You can simplify and make it more easier to read with following changes if you wish. At least consider

```ts
const initialState: Calendar = {

  state: "LOADING",

  guidanceEvents: [],

};



/**

 * Reducer function for calendar

 *

 * @param reducerState reducerState

 * @param action action

 * @returns returns state

 */

export const calendarReducer = (

  reducerState = initialState,

  action: ActionType

) => {

  switch (action.type) {

    case "UPDATE_CALENDAR_EVENTS_STATUS": {

      const newState: EventsState = action.payload;

      return {

        ...reducerState,

        state: newState,

      };

    }

    case "LOAD_CALENDAR_GUIDANCE_EVENTS": {

      const newEvents: CalendarEvent[] = action.payload;

      const newEventsFiltered = newEvents.filter((event) =>

        evaluateEvents(event, reducerState.guidanceEvents)

      );



      return {

        ...reducerState,

        guidanceEvents: [...reducerState.guidanceEvents, ...newEventsFiltered],

      };

    }

    case "ADD_CALENDAR_GUIDANCE_EVENT": {

      const newEvent: CalendarEvent[] = [action.payload];

      return {

        ...reducerState,

        guidanceEvents: [...reducerState.guidanceEvents, ...newEvent],

      };

    }

    case "UPDATE_CALENDAR_GUIDANCE_EVENT": {

      const updateEvent: CalendarEvent[] = [action.payload];

      const filteredState = reducerState.guidanceEvents.filter(

        (event) => event.id !== action.payload.id

      );



      return {

        ...reducerState,

        guidanceEvents: [...filteredState, ...updateEvent],

      };

    }

    case "DELETE_CALENDAR_GUIDANCE_EVENT": {

      if (typeof action.payload === "number") {

        return {

          ...reducerState,

          guidanceEvents: reducerState.guidanceEvents.filter(

            (event: CalendarEvent) => event.id !== (action.payload as number)

          ),

        };

      }

      const guidanceEvents = reducerState.guidanceEvents;

      const currentEventObject = action.payload as CalendarEvent;

      const currentEventIndex = guidanceEvents.findIndex(

        (event) => event.id === currentEventObject.id

      );

      guidanceEvents[currentEventIndex] = currentEventObject;



      return {

        ...reducerState,

        guidanceEvents: guidanceEvents,

      };

    }

    default:

      return reducerState;

  }
```